pisa/stages/xsec/dis_sys.py

```diff
@@ -0,0 +1,213 @@
+        )
+
+        assert self.input_mode is not None
+        assert self.calc_mode is None
+        assert self.output_mode is not None
+
+        self.extrapolation_type = extrapolation_type
+
+    @profile
+    def setup_function(self):
+
+        extrap_dict = from_file('cross_sections/tot_xsec_corr_Q2min1_isoscalar.pckl')
+
+        # load splines
+        wf_nucc =    from_file('cross_sections/dis_csms_splines_flat/NuMu_CC_flat.pckl')
+        wf_nubarcc = from_file('cross_sections/dis_csms_splines_flat/NuMu_Bar_CC_flat.pckl')
+        wf_nunc =    from_file('cross_sections/dis_csms_splines_flat/NuMu_NC_flat.pckl')
+        wf_nubarnc = from_file('cross_sections/dis_csms_splines_flat/NuMu_Bar_NC_flat.pckl')
+
+        # set this to events mode, as we need the per-event info to calculate these weights
+        self.data.data_specs = 'events'
+
+        for container in self.data:
+
+            # creat keys for external dict
+
+            if container.name.endswith('_cc'):
+                current = 'CC'
+            elif container.name.endswith('_nc'):
+                current = 'NC'
+            else:
+                raise ValueError('Can not determine whether container with name "%s" is pf type CC or NC based on its name'%container.name)
+            nu = 'Nu' if container['nubar'] > 0 else 'NuBar'
+
+            lgE = np.log10(container['true_energy'].get('host'))
+            bjorken_y = container['bjorken_y'].get('host')
+            dis = container['dis'].get('host')
+
+            #
+            # Calculate variation of total cross section
+            #
+
+            lgE_min = 2. if self.extrapolation_type == 'constant' else 1.68
+            mask = lgE <= lgE_min
+
+            w_tot = np.ones_like(lgE)
+            
+            poly_coef = extrap_dict[nu][current]['poly_coef']
+            w_tot[~mask] = np.polyval(poly_coef, lgE[~mask])
+           
+            if self.extrapolation_type == 'constant':
+                poly_coef = extrap_dict[nu][current]['poly_coef']
+                w_tot[mask] = np.polyval(poly_coef, lgE_min * w_tot[mask]) 
+            elif self.extrapolation_type == 'linear':
+                lin_coef = extrap_dict[nu][current]['linear']
+                w_tot[mask] = np.polyval(lin_coef, lgE[mask])
+            elif self.extrapolation_type == 'higher':
+                poly_coef = extrap_dict[nu][current]['poly_coef']
+                w_tot[mask] = np.polyval(poly_coef, lgE[mask]) 
+            else:
+                raise ValueError('Unknown extrapolation type "%s"'%self.extrapolation_type)
+
+            # mke centered arround 0, and set to 0 for all non-DIS events
+            w_tot -= 1
+            w_tot *= dis
+          
+            container["dis_correction_total"] = w_tot
+
+            #
+            # Calculate variation of differential cross section
+            #
+
+            lgE_min = 2.
+            mask = lgE <= lgE_min
+
+            w_diff = np.ones_like(lgE)
```

Comment:
empty_like

Reply:
the ones are needed later on

Reply:
Ah I see, you're multiplying N 1's by `lgE_min` to obtain N copies of `lgE_min`. Broadcasting applies just the same here as in numpy, so all you have to do is pass `lgE_min` as the first arg; below can become simply:
```python
w_diff = np.empty_like(lgE)
w_diff[valid_mask] = weight_func.ev(lgE[valid_mask], bjorken_y[valid_mask])
w_diff[extrapolation_mask] = weight_func.ev(lgE_min, bjorken_y[extrapolation_mask])
```
This avoids creation of an extra array and a bunch of multiplies, but since it's in the setup function that might not matter. Change if you want, then feel free to merge.